File: waiting_games/games/snake.py
"""Snake: steer your snake, eat the apples, don't crash. Last snake alive wins.

The first real-time game. There are no turns: every player may steer on every
tick, and a move is INTENT ("I am now heading left") that the clock acts on -- it
does not move the snake by itself. That is what stops a held arrow key from
turning into a message per frame.

Solo is a first-class mode (min_players = 1): survive as long as you can, and
your length is the score.

-- why a move is DECIDED a tick before it HAPPENS ---------------------------

A tick used to read the intent standing at the instant it fired, and move. That
is the obvious design and it makes the game unsteerable, for a reason that has
nothing to do with the rules and everything to do with the fact that a snake is
drawn moving and the server moves it in jumps.

The browser has to slide the snake between cells or it lurches a whole square six
times a second. It can only slide between two cells it knows -- so it has to wait
for the far one, which means drawing the snake a cell behind where the server has
already put it. That cell is the lag: you steer the snake you can SEE, and it is
not the one the server has. Slowing the game down cannot help, because the delay
is one tick and a tick IS one cell, at every possible tick rate.

So decide each move one tick early and say so. `step` is what a snake WILL do on
the next tick -- the cell it takes, or nothing if that move kills it -- fixed at
the end of the tick before, and sent out with the state. The browser is then
animating a move that has already been decided rather than guessing at one, so it
can draw it AS IT HAPPENS, in real time, with no cell of delay and nothing to
snap back from when the guess was wrong.

What it costs is that your key lands on the move after the one already in flight.
That is not new latency -- it is the same tick you always waited, moved to where
it can be seen -- and it nets out FASTER, because the screen stops being a cell
behind. Half a tick to see your own turn, where it used to be a tick and a half.
"""

from __future__ import annotations

import random
from dataclasses import dataclass

from .base import InvalidMove, RealTimeGame, Result

WIDTH = HEIGHT = 24
START_LENGTH = 3
APPLES = 3
GROWTH = 2  # cells gained per apple

DIRECTIONS = {
    "up": (0, -1),
    "down": (0, 1),
    "left": (-1, 0),
    "right": (1, 0),
}


@dataclass
class Step:
    """What a snake does on the NEXT tick. Decided at the end of this one.

    This is the whole contract with the renderer: it is a fact about a tick that
    has not happened yet, which is what lets the browser draw the move while it is
    happening instead of a cell after it did.
    """

    heading: str
    to: tuple[int, int] | None  # the cell the head takes; None if the move kills it
    grows: bool  # ...and whether the tail stays put, because it ate

    @property
    def fatal(self) -> bool:
        return self.to is None


class SnakeBody:
    def __init__(self, cells: list[tuple[int, int]], heading: str) -> None:
        self.cells = cells  # head first
        self.heading = heading  # the direction of the move it last made
        self.pending = heading  # ...and the one the player has asked for since
        self.growing = 0
        self.alive = True
        self.step: Step | None = None  # what it does next: set by _decide()

    @property
    def head(self) -> tuple[int, int]:
        return self.cells[0]

    @property
    def facing(self) -> str:
        """The direction it will be travelling once its next move is made.

        Not the one it last travelled. By the time a fresh intent can be acted on,
        the move already in flight has happened -- so this is the direction the
        neck will be in, and the direction a reversal has to be judged against.
        """
        return self.step.heading if self.step else self.heading


class Snake(RealTimeGame):
    key = "snake"
    title = "Snake"
    min_players = 1
    max_players = 6
    # The snake moves one cell per tick, so this IS the speed -- there is no
    # separate difficulty knob, and there must not be one. It was 8 Hz, which
    # players found too fast to steer: a 24-cell board crossed in three seconds
    # leaves about a sixth of a second to see a wall coming and do something
    # about it. 6 Hz gives you a third.
    #
    # It is on the wire (public_state), and the renderer slides the snake across
    # exactly this interval, so tuning it here is the whole change.
    tick_hz = 6.0

    def __init__(self) -> None:
        super().__init__()
        self.snakes: list[SnakeBody] = []
        self.apples: list[tuple[int, int]] = []
        self.rng = random.Random()
        self.elapsed = 0.0

    def _on_start(self) -> None:
        """Space the snakes evenly down the board, all facing right."""
        self.snakes = []
        for seat in range(len(self.players)):
            row = (seat + 1) * HEIGHT // (len(self.players) + 1)
            cells = [(START_LENGTH - i, row) for i in range(START_LENGTH)]
            self.snakes.append(SnakeBody(cells, "right"))

        self.apples = []
        for _ in range(APPLES):
            self._grow_an_apple()

        # The first tick executes a move like every other one, so there has to be
        # a move for it to execute.
        self._decide()

    # -- the world -------------------------------------------------------

    def _occupied(self) -> set[tuple[int, int]]:
        taken = {cell for snake in self.snakes if snake.alive for cell in snake.cells}
        taken.update(self.apples)
        return taken

    def _grow_an_apple(self) -> None:
        # _occupied() is built ONCE. It used to be called from inside the filter,
        # which rebuilt the whole occupied set for every one of the board's 576
        # cells -- and the set is every cell of every snake, so the cost climbed
        # with the snake: 0.3ms when it was three cells long, 4ms at eighty.
        #
        # That lands on the tick that eats the apple, inside the tick loop, and it
        # pushed that one tick ~12ms late. The browser slides the snake between
        # ticks and measures how long it has to slide for, so a late tick made it
        # stall and then jump -- once every few seconds, which is how often you eat.
        # Smooth animation is a bargain the server has to keep too.
        taken = self._occupied()
        free = [
            (x, y) for x in range(WIDTH) for y in range(HEIGHT) if (x, y) not in taken
        ]
        if free:
            self.apples.append(self.rng.choice(free))

    # -- input is intent, not action --------------------------------------

    def _apply(self, seat: int, move: dict) -> None:
        heading = move.get("dir")
        if heading not in DIRECTIONS:
            raise InvalidMove("snake.unknown_direction")

        snake = self.snakes[seat]
        if not snake.alive:
            raise InvalidMove("snake.dead")

        # No reversing into your own neck. Judged against the direction the snake
        # will be FACING once the move already in flight has been made -- see
        # SnakeBody.facing -- because that is the move this intent follows. It used
        # to be judged against the direction last travelled, which was the same
        # thing back when a tick read the intent standing at the moment it fired.
        dx, dy = DIRECTIONS[heading]
        cx, cy = DIRECTIONS[snake.facing]
        if len(snake.cells) > 1 and (dx, dy) == (-cx, -cy):
            raise InvalidMove("snake.no_reverse")

        snake.pending = heading  # _decide() is what actually turns the snake

    # -- the clock --------------------------------------------------------

    def tick(self, dt: float) -> None:
        self.elapsed += dt

        if not any(snake.alive for snake in self.snakes):
            return

        self._execute()  # ...the move decided last tick, which is on every screen
        self._settle()
        self._decide()  # ...and the next one, from the intent standing right now

    def _execute(self) -> None:
        """Carry out the move every client has already been told about."""
        for snake in self.snakes:
            if not snake.alive or snake.step is None:
                continue

            if snake.step.fatal:
                snake.alive = False
                continue

            head = snake.step.to
            snake.heading = snake.step.heading
            snake.cells.insert(0, head)

            if head in self.apples:
                self.apples.remove(head)
                snake.growing += GROWTH
                self._grow_an_apple()

            if snake.growing:
                snake.growing -= 1  # keep the tail: the snake gets longer
            else:
                snake.cells.pop()

    def _decide(self) -> None:
        """Fix what happens on the NEXT tick, and hand it to the renderer.

        Everything the old tick() worked out at the moment of moving is worked out
        here instead, one tick earlier. It is the same arithmetic on the same
        board -- the only thing that has moved is WHEN, and the point of moving it
        is that the answer can now be put on the wire before it is needed.
        """
        living = [snake for snake in self.snakes if snake.alive]

        # Everyone moves at once, so work out every new head BEFORE resolving any
        # collision. Doing it snake-by-snake would let whoever is first in the
        # list win a head-on that should kill them both.
        heads: dict[SnakeBody, tuple[int, int]] = {}
        for snake in living:
            dx, dy = DIRECTIONS[snake.pending]
            x, y = snake.head
            heads[snake] = (x + dx, y + dy)

        # A tail cell moves out of the way next tick, so following one is legal --
        # unless its owner is growing, in which case the tail stays put.
        bodies: set[tuple[int, int]] = set()
        for snake in living:
            keep = snake.cells if snake.growing else snake.cells[:-1]
            bodies.update(keep)

        doomed: set[SnakeBody] = set()
        for snake, head in heads.items():
            x, y = head
            if not (0 <= x < WIDTH and 0 <= y < HEIGHT):
                doomed.add(snake)  # into the wall
            elif head in bodies:
                doomed.add(snake)  # into somebody, possibly themselves
            else:
                # Two heads into the same cell: both of them.
                rivals = [s for s, h in heads.items() if s is not snake and h == head]
                if rivals:
                    doomed.add(snake)
                    doomed.update(rivals)

        for snake in living:
            head = heads[snake]
            snake.step = Step(
                heading=snake.pending,
                to=None if snake in doomed else head,
                # The tail holds if the snake is still growing from the last apple,
                # or is about to eat the next one. Without this the browser slides
                # the tail forward on a snake that is getting longer, and the state
                # puts it back a sixth of a second later -- a tail that flickers
                # once per apple, every apple.
                grows=snake.growing > 0 or head in self.apples,
            )

    def _settle(self) -> None:
        alive = [seat for seat, snake in enumerate(self.snakes) if snake.alive]
        solo = len(self.players) == 1

        if solo:
            if not alive:
                self.finish(Result.draw())  # nobody to beat; the score is the length
            return

        if len(alive) == 1:
            self.finish(Result(winner_seat=alive[0]))
        elif not alive:
            self.finish(Result.draw())  # everybody crashed on the same tick

    # -- what everyone sees -----------------------------------------------

    def public_state(self) -> dict:
        return {
            "width": WIDTH,
            "height": HEIGHT,
            "apples": [list(apple) for apple in self.apples],
            # `seconds` is the SCORE, not a clock to animate against: it is rounded
            # to a tenth, and at these rates a tenth cannot tell one tick from two. The
            # renderer uses the platform's `tick` and `tickHz` for that.
            "seconds": round(self.elapsed, 1),
            "snakes": [
                {
                    "player": self.players[seat],
                    "cells": [list(cell) for cell in snake.cells],
                    "alive": snake.alive,
                    "length": len(snake.cells),
                    # What it does NEXT tick, so the browser can draw the move
                    # while it happens instead of a cell after it did. `next` is
                    # the cell the head takes -- null if that move kills it, and
                    # then the snake simply stops, which is what the engine does
                    # too: it dies where it stands rather than in the wall.
                    #
                    # This is not hidden information. Snake has none: every cell of
                    # every snake is already on the wire, and a move that is fixed
                    # is a fact about the board, not a peek at somebody's keyboard.
                    "next": (
                        list(snake.step.to)
                        if snake.alive and snake.step and snake.step.to
                        else None
                    ),
                    "grows": bool(snake.alive and snake.step and snake.step.grows),
                }
                for seat, snake in enumerate(self.snakes)
            ],
        }

File: waiting_games/static/games/snake.js
// Snake.
//
// update() does NO dom work -- it just stashes the latest frame. The rAF loop
// paints from it. That is what keeps a 6 Hz state stream from stuttering a
// 60 Hz screen, and what stops the canvas being rebuilt on every push.

import { t } from "../i18n.js";
import { COLOURS, canvas, hint, keys, onChange, swipe } from "./_canvas.js";
import { readout } from "./_debug.js";
import { ahead, at, timeline } from "./_interpolate.js";
import * as rules from "./snake_rules.js";

// Swipe on a phone; the arrow keys (or WASD) on anything with them.
const SWIPES = {
  up: { dir: "up" },
  down: { dir: "down" },
  left: { dir: "left" },
  right: { dir: "right" },
};

const ARROWS = {
  ArrowUp: { dir: "up" },
  ArrowDown: { dir: "down" },
  ArrowLeft: { dir: "left" },
  ArrowRight: { dir: "right" },
  w: { dir: "up" },
  s: { dir: "down" },
  a: { dir: "left" },
  d: { dir: "right" },
};

// A solo game, played HERE.
//
// The server deals the board, hands over the seed, and does not tick. We run the
// rules ourselves -- snake_rules.js, the twin of snake.py, pinned to it by
// tests/test_determinism.py -- and report the run when it ends, for the server to
// play again and check.
//
// This exists because a phone cannot be shown this game over a radio. Measured, on
// a real one, mid-game: the gaps between state arrivals ran a median of 168ms (the
// tick, exactly), a 90th percentile of 333ms, and a worst of 2439ms, with nothing
// dropped. Every state was sent on time and then held in the air until the radio
// woke up. There is no renderer that survives that, and three of them were written
// trying: you can be smooth or you can be current, and on a line that stalls for
// two seconds you cannot be either for long.
//
// So take the packet out of the render loop. Your turn lands on the next LOCAL
// tick, which is the floor -- a snake cannot turn in the middle of a cell -- and
// the radio may sleep for as long as it likes.
function local(game, me, publish, finish) {
  const players = game.snakes.map((snake) => snake.player);
  const board = rules.game(game.seed, players);
  board.start();

  const seat = players.indexOf(me.sub);
  const tickMs = 1000 / game.tickHz;
  const moves = []; // the input log: what the server will replay to check us
  let spare = 0; // time owed to the simulation but not yet ticked
  let painted = null;
  let reported = false;

  const show = () => {
    const state = board.state();
    publish?.({
      ...state,
      status: board.over ? "finished" : "active",
      over: board.over,
      // Solo ends as a DRAW on the wire -- Result.draw(), because there was nobody
      // to beat. snake.js's outcome() is what turns that into the loss it actually
      // was, and it reads these.
      draw: board.over,
      winner: null,
    });
  };
  show();

  return {
    seat,

    /** Advance to `now` and hand back the moment to draw: a state, and how far
     * through the move in flight we are. A fixed timestep, so the simulation runs
     * the same whatever the screen is doing. */
    read(now) {
      if (painted === null) painted = now;
      // A backgrounded tab hands back a gap of many seconds, and simulating a
      // minute of Snake in one frame is not catching up, it is a death sentence
      // delivered while you were not looking. Cap it, and let the clock slip.
      spare += Math.min(now - painted, 500);
      painted = now;

      while (spare >= tickMs && !board.over) {
        board.tick(1 / game.tickHz);
        spare -= tickMs;
        show();
      }

      if (board.over) {
        if (!reported) {
          reported = true;
          finish?.({ ticks: board.ticks, moves });
        }
        spare = 0;
      }

      return {
        state: board.state(),
        alpha: board.over ? 1 : Math.min(spare / tickMs, 1),
        ahead: true,
      };
    },

    /** Steer, right now, on the next tick this simulation runs. Returns whether
     * the rules allowed it -- a reversal into your own neck is refused here by the
     * same code that refuses it on the server. */
    steer(dir) {
      if (board.over || !board.steer(seat, dir)) return false;
      // The tick it lands on, which is the next one to run. run() in snake.py
      // applies a move at tick k and THEN ticks k, so these line up exactly.
      moves.push({ tick: board.ticks, dir });
      return true;
    },
  };
}

export function create({ root, me, send, publish, finish }) {
  // Snake moves ONE WHOLE CELL per tick. Drawn at the cell it is on, it therefore
  // teleports a full grid square six times a second, and no amount of repainting
  // at 60 fps helps: the thing being painted only changes 6 times a second. That
  // is the choppiness, and it is not a dropped-frame problem -- it is what the
  // game looks like when you draw it honestly.
  //
  // So slide it between cells. That used to mean drawing the world a tick in the
  // past, because a slide needs the cell it is going TO and only the server knew
  // it -- and a tick in the past is a whole CELL in the past, at any tick rate.
  // Now the server sends the cell each snake is about to take (`next`), so the
  // slide is drawn while the move is happening rather than after it happened.
  // Nothing here is a guess, and nothing is a cell behind.
  let clock = null;

  // When the browser owns the clock -- see local() above -- this is the game, and
  // the timeline above is not used at all. There is nothing to interpolate towards:
  // we ARE the server.
  let mine = null;
  let tickMs = 1000 / 6; // ...until the first state says otherwise

  const board = canvas(root, (context, side) => {
    const now = performance.now();
    const frame = mine ? mine.read(now) : clock?.read(now);
    if (!frame) return;

    paint(context, side, frame, me);

    // Nothing unless the page was loaded with ?debug. It reads the head off the
    // very cells that were just painted, so what it reports is what is on the
    // screen and not a second opinion about it.
    if (meter) {
      const ours = frame.state.snakes.find((snake) => snake.player === me.sub);

      // A game we are running ourselves has no connection to report on: there is
      // no buffer, no delay, and no state that can be late, because there is no
      // state coming. That is the entire point of it, and the readout should say
      // so rather than divide by a timeline that does not exist.
      const stats = clock
        ? clock.debug(now)
        : { delayTicks: 0, tickMs, dropped: 0, stateAge: 0 };

      if (ours) {
        meter.frame({
          now,
          head: where(frame, ours, 0),
          path: mine ? "local" : frame.ahead ? "live" : "catching up",
          delayMs: stats.delayTicks * stats.tickMs,
          tickMs: stats.tickMs,
          dropped: stats.dropped,
          stateAge: stats.stateAge,
        });
      }
    }
  });
  hint(root, t("snake.hint"));

  // After canvas(), which replaces the root's children -- same reason hint() is.
  // The paint loop above closes over this, and only ever runs from a
  // requestAnimationFrame, which is to say after this line.
  //
  // It gets the board so it can count the raw pointer events the browser delivers,
  // BEFORE swipe() has an opinion about them. "The touch controls did not work" is
  // several different bugs -- a listener that never fired, a gesture iOS took back
  // to scroll with, a swipe below the threshold, a move the server rejected -- and
  // they are indistinguishable from the sofa and need different fixes.
  const meter = readout(root, board.element);

  // One deduper for every way of steering. A swipe, a key and a button are all
  // the same intent, and if each kept its own idea of what the server was last
  // told they would disagree the moment somebody used two of them.
  const post = onChange(send);
  const intend = (input) => {
    // A game we are running ourselves needs no wire and no dedupe: the answer to
    // "did that turn land" is yes, on the next local tick, and the rules that
    // would have refused it refuse it here instead. Nothing goes out until the
    // run is over, and then the whole log goes at once.
    if (mine) {
      if (mine.steer(input.dir)) meter?.pressed();
      return;
    }

    // Only time a press that actually goes out. onChange() swallows a repeat --
    // you pressed the way you were already going -- and a swallowed press has no
    // turn coming to answer it, so a stopwatch started on one would run until the
    // NEXT real turn and report that wait as if it were this press's.
    if (post(input)) meter?.pressed();
  };
  const stopKeys = keys(ARROWS, intend);
  const stopSwipe = swipe(board.element, intend, SWIPES);

  return {
    update(game) {
      // Before the clock starts there is nothing to animate -- and nothing to
      // animate BETWEEN, either: a game sitting in the lobby has no ticks.
      if (game.tick === undefined) return;

      // OUR game, or theirs? A solo board is dealt by the server and then played
      // here; anything with a second snake in it is arbitrated there, and we watch.
      tickMs = 1000 / game.tickHz;

      if (game.clientClock) {
        if (mine === null && game.status === "active" && game.snakes.length) {
          mine = local(game, me, publish, finish);
        }
        return; // the server is not ticking this one. There is nothing to accept.
      }

      if (clock === null) clock = timeline(game.tickHz);

      // The tick index is what makes a state a MOMENT rather than just the
      // newest thing we have. A push that carries no new tick -- somebody
      // connecting, somebody leaving -- is dropped on the floor by accept(),
      // which is why opening the page no longer stutters everyone else's snake.
      const now = performance.now();
      clock.accept(game, game.tick, now);

      // The gaps between arrivals are the connection's real character, and they
      // are what every constant in _interpolate.js is tuned against. Measure them
      // on the machine that has the problem.
      meter?.arrived(now, game.tick);
    },
    destroy() {
      board.destroy();
      stopKeys();
      stopSwipe();
    },
  };
}

// Forwards into the move the server says is coming, or -- when a state went
// missing and we are behind -- backwards down the body towards the state we did
// get. A dead snake is not going anywhere at all: it dies WITHOUT moving, so its
// body still reads as a perfectly good path, and walking it in either direction
// would drag the corpse off the cell it died on. Both walks refuse.
//
// The readout calls this too, on the same frame, so the speed it reports is the
// speed of the snake that was actually drawn rather than a second opinion.
function where(frame, snake, index) {
  return frame.ahead
    ? ahead(snake, index, frame.alpha)
    : at(snake, index, frame.alpha, snake.alive ? frame.span : 0);
}

function paint(context, side, frame, me) {
  const game = frame.state;
  const cell = side / game.width;

  context.fillStyle = "rgba(127, 140, 160, 0.10)";
  context.fillRect(0, 0, side, cell * game.height);

  for (const [x, y] of game.apples) {
    context.fillStyle = "#d9534f";
    context.beginPath();
    context.arc((x + 0.5) * cell, (y + 0.5) * cell, cell * 0.34, 0, Math.PI * 2);
    context.fill();
  }

  game.snakes.forEach((snake, seat) => {
    context.fillStyle = COLOURS[seat % COLOURS.length];
    context.globalAlpha = snake.alive ? 1 : 0.25;

    snake.cells.forEach((_, index) => {
      const [x, y] = where(frame, snake, index);
      const inset = index === 0 ? 0 : cell * 0.08; // the head is a little bigger
      context.fillRect(
        x * cell + inset,
        y * cell + inset,
        cell - inset * 2,
        cell - inset * 2,
      );
    });

    // Ring your own head so you can find yourself in a six-snake scrum.
    if (snake.player === me.sub && snake.alive) {
      const [hx, hy] = where(frame, snake, 0);
      context.strokeStyle = "#fff";
      context.lineWidth = 2;
      context.strokeRect(hx * cell + 1, hy * cell + 1, cell - 2, cell - 2);
    }

    context.globalAlpha = 1;
  });
}

export function outcome(game, me) {
  // Snake has no winner to be. The run ends when you crash, and the engine reports
  // that as a DRAW -- Result.draw() -- because there was nobody to beat. Which is
  // correct on the wire and would be absurd in the room: the platform's rule would
  // play a neutral, that-was-close chime at somebody who just died.
  //
  // You crashed. It is a loss. Have the trombone. A watcher gets neither.
  return game.snakes[0]?.player === me.sub ? "lose" : "none";
}

export function describe(game, me) {
  // Not before the board is dealt. This is deliberately NOT `!== "active"`:
  // a finished session's status is "finished", so that test quietly swallowed
  // every game-over line below it and left the platform's generic one to speak
  // for a game it does not understand.
  if (game.status === "waiting") return null;

  // A spectator has no snake of their own -- any logged-in user may open a game
  // socket and watch -- so nothing here may reach through `mine` without checking.
  // Watching somebody else's run used to throw and blank the whole status line.
  const mine = game.snakes.find((snake) => snake.player === me.sub);

  if (game.over) {
    return mine
      ? t("snake.solo_over", { seconds: game.seconds, length: mine.length })
      : t("snake.solo_over_watched", { seconds: game.seconds });
  }

  // No clock on this line. The server streams `seconds` to one decimal place,
  // once per tick, and a status line that rewrites itself several times a second
  // is one nobody can actually read -- the digits just flicker. The elapsed time
  // IS the score, so it is reported once, when the run ends.
  if (!mine) return t("snake.watching");
  return mine.alive
    ? t("snake.length", { length: mine.length })
    : t("snake.you_crashed");
}
